### ISSUE TITLE
[FEATURE] Guest cluster nodes distributed across failure domains

### ISSUE THREAD

**Issue by janeczku** [harvester/harvester#2316]
**Problem:**
The VMs of guest clusters provisioned with the Harvester node driver may end up being scheduled to a single physical server. This is against the Kubernetes best practice to distribute cluster nodes across multiple failure domains to improve system availability. 

**Solution:**

When provisioning guest clusters on Harvester, the user can specify for each machine pool a list of topology keys (matching labels assigned to the physical Harvester servers) that will be used to distribute VM nodes across failure domains.

Example topology keys:

- `topology.kubernetes.io/zone`
- `kubernetes.io/hostname`
- `rack`

**Implementation detail:**

[Pod topology spread constraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/) may be used to control how the VirtualMachineInstances of the guest cluster are spread across the user defined failure domains.

**Comment by chrisho** [comment]
> [Pod topology spread constraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/) may be used to control how the VirtualMachineInstances of the guest cluster are spread across the user defined failure domains.

The latest version of kubevirt does not support topology spread constrains, currently only through PodAffinity/PodAntiffinity to solve the problem, but, do you have any ideas? What is the user-specific scenario?

**Comment by DaiYuzeng** [comment]
@chrisho This issue will be ready to test after  https://github.com/rancher/dashboard/issues/6366 is fixed, thanks!

**Comment by TachunLin** [comment]
Verified **fixed** on Harvester`master-08d01323-head` (07/19) with Rancher `v2.6.7-rc3` Close this issue.

In this feature we would focus on checking the `Node scheduling` selector and `Pod scheduling` selector can deploy guest VM to the Harvester node and pod matching the label. 

### Test Plan

We have the following test plan:
1. Check **can** deploy guest VM to the specific Harvester node matching the node selector
2. Check **can't** (failed to) deploy guest VM to the specific node if any of the node selector mismatched 

We use the following ui-index url and node driver for testing:

**Rancher UI settings:** 
- ui-index: https://yuzeng-dev.oss-cn-beijing.aliyuncs.com/node-driver/latest2/index.html (for RKE1)
- ui-dashboard-index: https://releases.rancher.com/harvester-ui/dashboard/latest/index.html (for RKE2)
- ui-offline-preferred: Remote

**Node driver:**
- https://releases.rancher.com/harvester-node-driver/v0.5.0/docker-machine-driver-harvester-amd64.tar.gz

### Test Result

- [x] **Can** deploy guest RKE1 cluster vm to specific Harvester node matching the `node scheduling` and `pod scheduling` selector
- [x] **Can** deploy guest RKE2 cluster vm to specific Harvester node matching the `node scheduling` and `pod scheduling` selector
- [x] **Can't** deploy guest RKE1 cluster vm to specific Harvester node if any of the `node scheduling` and `pod scheduling` selector not found on the Harvester node
- [x] **Can't** deploy guest RKE2 cluster vm to specific Harvester node if any of the `node scheduling` and `pod scheduling` selector not found on the Harvester node

### Test Information
* Test Environment: 1 node harvester on local kvm machine
* Harvester version: master-08d01323-head (07/19)
* Rancher version: v2.6.7-rc3

### Environment Setup 
1. Prepare Harvester master node 
2. Prepare Rancher v2.6.7-rc3
3. Import Harvester to Rancher 
4. Set the experimental `ui-index` in Rancher global settings:
https://yuzeng-dev.oss-cn-beijing.aliyuncs.com/node-driver/latest2/index.html
  ![image](https://user-images.githubusercontent.com/29251855/179678362-06046277-0cc5-4a23-aec6-415d1335bdf7.png)
5. Set the experimental `ui-dashboard-index` in Rancher global settings:
https://releases.rancher.com/harvester-ui/dashboard/latest/index.html
  ![image](https://user-images.githubusercontent.com/29251855/179717212-47addd08-ed19-442f-b8e1-15bde04403a6.png)
  
6. Set ui-offline-preferred: Remote
  ![image](https://user-images.githubusercontent.com/29251855/179678535-a649167d-36b9-4cbf-b7f0-56242563f327.png)

7. Edit the Harvester node driver API
  ![image](https://user-images.githubusercontent.com/29251855/179678896-5accbb10-5b21-4b72-b609-39536435e1b4.png)
8. Unckeck the built in 
9. Add the checksum value
10. Update the url to https://releases.rancher.com/harvester-node-driver/v0.5.0/docker-machine-driver-harvester-amd64.tar.gz
  ![image](https://user-images.githubusercontent.com/29251855/179715097-3c3c665c-98ba-48a0-bb92-4a24fe432d7e.png)
  
11. Show and send request
12. Downloading node driver
  ![image](https://user-images.githubusercontent.com/29251855/179715694-cc2f863e-2316-43aa-b865-afaeb88c74d7.png)
13. Check node driver version
  ![image](https://user-images.githubusercontent.com/29251855/179716177-5cc5ab04-93d5-4e60-9288-e69facc80d3e.png)

### RKE2 Verification Steps
1. Open Harvester Host page then edit host config 
2. Add the following key value in the labels page:
  - topology.kubernetes.io/zone: zone1
  - topology.kubernetes.io/region: region1 
  ![image](https://user-images.githubusercontent.com/29251855/179680225-ac260aa2-e879-486a-9af4-478eeac1123d.png)

3. Create a virtual machine on Harvester 
  ![image](https://user-images.githubusercontent.com/29251855/179710444-243efb42-c2bf-48bc-ac4d-30b57d44dabe.png)
4. Navigate to https://[harvester vip]/c/local
5. Open the workload -> pod page
6. Edit the yaml of the vm pod 
  ![image](https://user-images.githubusercontent.com/29251855/179717764-1f64cd56-a8c0-4dc9-8cb7-cccb4de60033.png)
7.  Copy the vm pod lables  `harvesterhci.io/vmName: test-vm`
![image](https://user-images.githubusercontent.com/29251855/179718046-ce8a9df4-211c-48bb-b467-e8b7cd8d7eb7.png)

8. Open the RKE2 provisioning page 
9. Expand the show advanced 
10. Click add Node selector in `Node scheduling`
11. Use default `Required` priority
12. Click Add Rule
13. Provide the following key/value pairs
  - `topology.kubernetes.io/zone: zone1`
  - `topology.kubernetes.io/region: region1`
  ![image](https://user-images.githubusercontent.com/29251855/179719766-e5fba341-dbf5-4248-a18a-4c9bf3191756.png)

15. Add Node selector in `Pod scheduling`
16. Use default `Required` priority and `Affinity` type
17. Click Add Rule
18. Provide the following key/value pairs
  -  `harvesterhci.io/vmName: test-vm`
19. Select `topology.kubernetes.io/zone` in the `Topology key`
  ![image](https://user-images.githubusercontent.com/29251855/179720462-996d5b17-5fee-4e46-8070-00df8f71dc52.png)
20. Provide the following user data 
  ```
  password: 123456
  chpasswd: { expire: False }
  ssh_pwauth: True
  ```
21. Create the RKE2 guest cluster
22. Go to Harvester Virtual Machine page
23. Edit yaml of the RKE2 guest cluster 
  ![image](https://user-images.githubusercontent.com/29251855/179721625-d10e0ecf-a032-4465-9b89-b4b0bd2a629c.png)
24. Check the node and pod affinity label have written into the yaml 
  ![image](https://user-images.githubusercontent.com/29251855/179721997-2a06433d-c8fb-4cd5-8747-01ccd02db966.png)
25. Check the guest cluster VM have no error message
26. Check can provision RKE2 cluster without error
![image](https://user-images.githubusercontent.com/29251855/179723020-7a1fd7ce-b4f7-41bf-8f72-bc7743a3e68b.png)

27. Provide mismatch or not exists node scheduling or pod scheduling selector key/value pair in step 13 or 18
28. Provision another RKE2 cluster 
29. Check VM should have error icon with message and automatically suspend the RKE2 provisioning

### RKE1 Verification Steps
1. Follow the steps 1 ~ 7 of the RKE2 verification section
2. Go to Rancher Cluster Management page, add the RKE1 node template
3. Click add Node selector in `Node scheduling`
4. Use default `Required` priority
5. Click Add Rule
6. Provide the following key/value pairs
  - `topology.kubernetes.io/zone: zone1`
  - `topology.kubernetes.io/region: region1`
  ![image](https://user-images.githubusercontent.com/29251855/179686902-3007bb87-22cb-4b72-a174-0499d7ad818b.png)
7. Add Node selector in `Pod scheduling`
8. Use default `Required` priority and `Affinity` type
9. Click Add Rule
10. Provide the following key/value pairs
  -  `harvesterhci.io/vmName: test-vm`
11. Select `topology.kubernetes.io/zone` in the `Topology key`
  ![image](https://user-images.githubusercontent.com/29251855/179685927-ac261b97-2c5a-4885-9fc0-394f24646c8c.png)
12. Create the RKE1 guest cluster
13. Go to Harvester Virtual Machine page
14. Edit yaml of the RKE1 guest cluster 
15. Check the node and pod affinity label have written into the yaml 
  ![image](https://user-images.githubusercontent.com/29251855/179690994-6f81d6e6-344e-4fb7-8ad2-1d7ff062aa05.png)
16. Check the guest cluster VM have no error message
17. Check can provision RKE1 cluster without error
  ![image](https://user-images.githubusercontent.com/29251855/179691115-15df5a8d-4ca4-49bf-943a-894b75aa77c5.png)
18. Provide mismatch or not exists node scheduling or pod scheduling selector key/value pair in step 6 and 10
19. Provision another RKE1 cluster 
20. Check VM should have error icon with message and automatically suspend the RKE1 provisioning

**Comment by DaiYuzeng** [comment]
Hi! @TachunLin 

We missed the case of `External Harvester Cluster`, can you please re-verify the aspect of `External Harvester Cluster` for this issue? Thanks!


These fields are not selectable, you can only input something since UI can not get related resources of these fields to afford users to select if credential type is `External`.

![image](https://user-images.githubusercontent.com/15041915/179919840-0153a692-0bde-4a07-aabc-c2e014420d56.png)

`External Credential`
![image](https://user-images.githubusercontent.com/15041915/179920549-4d7f2490-e517-4ffa-8b9a-467ed48ae9f2.png)

**Comment by TachunLin** [comment]
Verified **fixed** on Harvester `master-538d41fe-head` (08/03) with Rancher `v2.6.7-rc6` Close this issue.

The enhancement focus on using external cloud credential with Node scheduling` selector and `Pod scheduling` selector to deploy guest VM to specific node

### Test Result
- [x] **Can** deploy guest RKE1 cluster vm to specific Harvester node matching the `node scheduling` and `pod scheduling` selector using external Harvester cloud credential 
  ![image](https://user-images.githubusercontent.com/29251855/182821039-8348644c-dc76-4f5c-aa87-5a986cf6f130.png)
  ![image](https://user-images.githubusercontent.com/29251855/182822307-0e499af9-6930-4299-979e-5e13ee57ef7f.png)

- [x] **Can** deploy guest RKE2 cluster vm to specific Harvester node matching the `node scheduling` and `pod scheduling` selector using external Harvester cloud credential 
  ![image](https://user-images.githubusercontent.com/29251855/182817684-71a1bcdc-ede4-4637-924f-268e978ad9c8.png)

### Test resource 
We use the following ui-index url and node driver for testing:

**Rancher UI settings:** 
- ui-index: https://releases.rancher.com/ui/latest2/index.html (for RKE1)
- ui-dashboard-index: https://releases.rancher.com/dashboard/latest/index.html (for RKE2)
- ui-offline-preferred: Remote

**Node driver:**
- https://releases.rancher.com/harvester-node-driver/v0.5.0/docker-machine-driver-harvester-amd64.tar.gz

### Test Information
* Test Environment: 1 node harvester on local kvm machine
* Harvester version: `master-538d41fe-head` (08/03)
* Rancher version: v2.6.7-rc6

### Environment Setup 
1. Prepare Harvester master node 
2. Prepare Rancher v2.6.7-rc6
3. Import Harvester to Rancher 
4. Set the experimental `ui-index` in Rancher global settings:
5. Set the experimental `ui-dashboard-index` in Rancher global settings:
6. Set ui-offline-preferred: Remote
7. Go to Harvester Support page 
8. Download Kubeconfig 
9. Copy the content of Kubeconfig

### RKE2 Verification Steps
1. Open Harvester Host page then edit host config 
2. Add the following key value in the labels page:
  - topology.kubernetes.io/zone: zone_master
  - topology.kubernetes.io/region: region_master
  ![image](https://user-images.githubusercontent.com/29251855/182814590-50efebc3-701c-4d28-8fcf-be97ee66bfce.png)

3. Create a virtual machine on Harvester 
4. Navigate to https://[harvester vip]/dashboard/c/local
5. Open the workload -> pod page
6. Edit the yaml of the vm pod 
10. Copy the vm pod lables  `harvesterhci.io/vmName: test-vm`
11. Open the RKE2 provisioning page
12. Click the create the new cloud credential 
![image](https://user-images.githubusercontent.com/29251855/180426995-11527090-1e8f-4a6f-be95-7400822be26e.png)
13. Select external Harvester 
14. Paste the Harvester Kubeconfig content
  ![image](https://user-images.githubusercontent.com/29251855/180427873-bfebb49b-b1e4-426f-b224-e2d246984e97.png)
15. Expand the show advanced 
16. Click add Node selector in `Node scheduling`
17. Use default `Required` priority
18. Click Add Rule
19. Provide the following key/value pairs
  - `topology.kubernetes.io/zone: zone_master`
  - `topology.kubernetes.io/region: region_master`
20. Add Node selector in `Pod scheduling`
21. Use default `Required` priority and `Affinity` type
22. Click Add Rule
23. Provide the following key/value pairs
  -  `harvesterhci.io/vmName: test-vm`
24. Select `topology.kubernetes.io/zone` in the `Topology key`
25. Provide the following user data 
  ```
  password: 123456
  chpasswd: { expire: False }
  ssh_pwauth: True
  ```
26. Create the RKE2 guest cluster
27. Go to Harvester Virtual Machine page
28. Edit yaml of the RKE2 guest cluster 
29. Check the node and pod affinity label have written into the yaml 
  ![image](https://user-images.githubusercontent.com/29251855/180423959-f14e016e-7603-43d6-a2c9-f2a03ec0c8a0.png)
30. Check the guest cluster VM have no error message
31. Check can provision RKE2 cluster 


### RKE1 Verification Steps
1. Follow the steps 1 ~ 7 of the RKE2 verification section
2. Go to Rancher Cluster Management page, add the RKE1 node template
3. Select external cloud credential
4. Click add Node selector in `Node scheduling`
5. Use default `Required` priority
6. Click Add Rule
7. Provide the following key/value pairs
  - `topology.kubernetes.io/zone: zone_master`
  - `topology.kubernetes.io/region: region_master`
8. Add Node selector in `Pod scheduling`
9. Use default `Required` priority and `Affinity` type
10. Click Add Rule
11. Provide the following key/value pairs
  -  `harvesterhci.io/vmName: test-vm`
12. Select `topology.kubernetes.io/zone` in the `Topology key`
13. Create the RKE1 guest cluster
14. Go to Harvester Virtual Machine page
15. Edit yaml of the RKE1 guest cluster 
16. Check the node and pod affinity label have written into the yaml 
18. Check the guest cluster VM have no error message
19. Check can provision RKE1 cluster without error
